src/pool/srv_pool.c

```diff
@@ -3600,6 +3723,314 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
+{
+	d_iov_t			value;
+	uint64_t		val;
+	uint32_t		val32;
+	int			rc;
+	bool			need_commit = false;
+	uuid_t		       *hdl_uuids = NULL;
+	size_t			hdl_uuids_size;
+	int			n_hdl_uuids = 0;
+
+	if (rpc) {
+		rc = find_hdls_to_evict(tx, svc, &hdl_uuids, &hdl_uuids_size,
+					&n_hdl_uuids, NULL);
+		if (rc)
+			return rc;
+		D_DEBUG(DB_MGMT, "number of handles found was: %d\n", n_hdl_uuids);
+	}
+
+	if (n_hdl_uuids > 0) {
+		rc = pool_disconnect_hdls(tx, svc, hdl_uuids, n_hdl_uuids,
+					  rpc->cr_ctx);
+		if (rc != 0)
+			D_GOTO(out_free, rc);
+		need_commit = true;
+	}
+
+	d_iov_set(&value, NULL, 0);
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_policy,
+			   &value);
+	if (rc && rc != -DER_NONEXIST) {
+		return rc;
+	} else if (rc == -DER_NONEXIST) {
+		value.iov_buf = DAOS_PROP_POLICYSTR_DEFAULT;
+		value.iov_len = strlen(DAOS_PROP_POLICYSTR_DEFAULT);
+		value.iov_buf_len = value.iov_len;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_policy,
+				   &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade pool policy of pool "
+				"%d.\n", DP_UUID(pool_uuid), rc);
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	d_iov_set(&value, &val, sizeof(val));
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_redun_fac,
+			   &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST) {
+		val = DAOS_RPOP_PO_REDUN_FAC_DEFAULT;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_redun_fac, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade redundancy factor of pool, "
+				"%d.\n", DP_UUID(pool_uuid), rc);
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_ec_pda, &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST) {
+		val = DAOS_PROP_PO_EC_PDA_DEFAULT;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_ec_pda, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade EC performance domain "
+				"affinity of pool, %d.\n", DP_UUID(pool_uuid), rc);
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_rp_pda, &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST) {
+		val = DAOS_PROP_PO_RP_PDA_DEFAULT;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_rp_pda, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade RP performance domain "
+				"affinity of pool, %d.\n", DP_UUID(pool_uuid), rc);
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	d_iov_set(&value, &val32, sizeof(val32));
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_global_version, &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST || val32 < DAOS_META_GLOBAL_VERSION) {
+		val32 = DAOS_META_GLOBAL_VERSION;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_global_version, &value);
```

Comment:
Should we update ds_pool_prop_global_version only when the upgrade to GLOBAL_VERSION has completed? If we update it here, we no longer know what previous version is, especially after a system restart. Then how can we know what actions (i.e., the concept from https://daosio.atlassian.net/wiki/spaces/DC/pages/11092492289/Interoperability+up+down+grade) need to be taken in order to upgrade to GLOBAL_VERSION?